spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,298 @@
+      return Collections.emptyList();
+    }
+    if (startOffset.equals(endOffset)) {
+      return Collections.emptyList();
+    }
+
+    CloseableIterable<IndexedTask> tasks =
+        getChanges(startOffset.snapshotId(), startOffset.index(), startOffset.isStartingSnapshotId());
+    List<IndexedTask> pendingTasks = Lists.newArrayList();
+    for (IndexedTask t : tasks) {
+      if (t.snapshotId() == endOffset.snapshotId()) {
+        pendingTasks.add(t);
+        break;
+      } else {
+        pendingTasks.add(t);
+      }
+    }
+
+    CloseableIterable<FileScanTask> splitTasks = TableScanUtil.splitFiles(
+        CloseableIterable.combine(pendingTasks.stream().map(IndexedTask::task).collect(Collectors.toList()), tasks),
+        splitSize);
+    return Lists.newArrayList(
+        TableScanUtil.planTasks(splitTasks, splitSize, lookback, openFileCost));
+  }
+
+  private StreamingOffset calculateStartingOffset() {
+    StreamingOffset startingOffset;
+    if (startSnapshotId != null) {
+      startingOffset = new StreamingOffset(startSnapshotId, -1, true);
+    } else {
+      List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+      if (snapshotIds.isEmpty()) {
+        // there's no snapshot currently.
+        startingOffset = StreamingOffset.START_OFFSET;
+      } else {
+        startingOffset = new StreamingOffset(snapshotIds.get(snapshotIds.size() - 1), -1, true);
+      }
+    }
+
+    return startingOffset;
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    if (start.equals(StreamingOffset.START_OFFSET)) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    CloseableIterable<IndexedTask> pendingTasks = getChanges(start.snapshotId(), start.index(),
+        start.isStartingSnapshotId());
+    CloseableIterable<IndexedTask> rateLimitedTasks = rateLimit(pendingTasks, maxFilesPerTrigger);
+    IndexedTask last = Iterables.size(rateLimitedTasks) == 0 ? null : Iterables.getLast(rateLimitedTasks);
+
+    if (last == null) {
+      return start;
+    } else {
+      boolean isStarting = last.snapshotId() == start.snapshotId() && start.isStartingSnapshotId();
+      return new StreamingOffset(last.snapshotId(), last.index(), isStarting);
+    }
+  }
+
+  @VisibleForTesting
+  CloseableIterable<IndexedTask> getChanges(long snapshotId, int index, boolean isStarting) {
+    List<CloseableIterable<IndexedTask>> indexedTasks = Lists.newArrayList();
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
+
+    if (isStarting) {
+      CloseableIterable<FileScanTask> iter = sortByFile(buildTableScan().useSnapshot(snapshotId).planFiles());
+      List<IndexedTask> tasks = Streams.mapWithIndex(Streams.stream(iter),
+          (t, i) -> new IndexedTask(snapshotId, t, (int) i))
+          .collect(Collectors.toList());
+      indexedTasks.add(CloseableIterable.combine(tasks, iter));
```

Comment:
I think it would be worth adding `transformWithIndex` to `CloseableIterable`:

```java
  static <I, O> CloseableIterable<O> transformWithIndex(CloseableIterable<I> iterable, BiFunction<Long, I, O> transform) {
    Preconditions.checkNotNull(transform, "Cannot apply a null transform");

    return new CloseableIterable<O>() {
      @Override
      public void close() throws IOException {
        iterable.close();
      }

      @Override
      public Iterator<O> iterator() {
        return new Iterator<O>() {
          private final Iterator<I> inner = iterable.iterator();
          private long index = 0;

          @Override
          public boolean hasNext() {
            return inner.hasNext();
          }

          @Override
          public O next() {
            O result = transform.apply(index, inner.next());
            index += 1;
            return result;
          }
        };
      }
    };
  }
```

Then you could make a reusable function for this:

```java
  private static CloseableIterable<IndexedTask> indexTasks(CloseableIterable<FileScanTask> tasks, long snapshotId) {
    return CloseableIterable.transformWithIndex(tasks, (index, task) -> new IndexedTask(snapshotId, task, index));
  }
```

Reply:
Updated.